File: trees_and_graphs/deepest_leaves_sum.java
public int deepestLeavesSumDFS(TreeNode root)
{
    int deepestSum = 0, depth = 0, currDepth;
    Deque<Pair<TreeNode, Integer>> stack = new ArrayDeque();
    stack.push(new Pair(root, 0));

    while (!stack.isEmpty())
    {
        Pair<TreeNode, Integer> p = stack.pop();
        root = p.getKey();
        currDepth = p.getValue();

        if (root.left == null && root.right == null)
        {
            // If this leaf is the deepest one seen so far
            if (depth < currDepth)
            {
                // Start new sum
                deepestSum = root.val;

                // Note new depth
                depth = currDepth;
            }
            else if (depth == currDepth)
            {
                // If there were already leaves at this depth,
                // update existing sum
                deepestSum += root.val;
            }
        }
        else
        {
            if (root.right != null)
            {
                stack.push(new Pair(root.right, currDepth + 1));
            }

            if (root.left != null)
            {
                stack.push(new Pair(root.left, currDepth + 1));
            }
        }
    }

    return deepestSum;
}

public int deepestLeavesSumBFS(TreeNode root)
{
    ArrayDeque<TreeNode> nextLevel = new ArrayDeque();
    ArrayDeque<TreeNode> currLevel = new ArrayDeque();
    nextLevel.offer(root);

    while (!nextLevel.isEmpty())
    {
        // Prepare for the next level
        currLevel = nextLevel.clone();
        nextLevel.clear();

        for (TreeNode node: currLevel)
        {
            // Add child nodes of the current level
            // in the queue for the next level
            if (node.left != null)
            {
                nextLevel.offer(node.left);
            }

            if (node.right != null)
            {
                nextLevel.offer(node.right);
            }
        }
    }

    int deepestSum = 0;
    for (TreeNode node: currLevel)
    {
        deepestSum += node.val;
    }
    return deepestSum;
}

File: trees_and_graphs/deepest_leaves_sum.py
import treenode

from collections import deque

def deepestLeavesSumDFS(root: treenode.TreeNode) -> int:
    deepest_sum = depth = 0
    stack = [(root, 0)]

    while stack:
        node, curr_depth = stack.pop()

        if node.left is None and node.right is None:
            # If this leaf is the deepest one seen so far
            if depth < curr_depth:
                # Start new sum
                deepest_sum = node.val

                # Note new depth
                depth = curr_depth
            elif depth == curr_depth:
                # If there were already leaves at this depth,
                # update existing sum
                deepest_sum += node.val
        else:
            if node.right:
                stack.append((node.right, curr_depth + 1))

            if node.left:
                stack.append((node.left, curr_depth + 1))
    
    return deepest_sum

def deepestLeavesSumBFS(root: treenode.TreeNode) -> int:
    next_level = deque([root,])

    while next_level:
        # Prepare for the next level
        curr_level = next_level
        next_level = deque()

        for node in curr_level:
            # Add child nodes of the current level
            # in the queue for the next level
            if node.left:
                next_level.append(node.left)
            
            if node.right:
                next_level.append(node.right)
        
    return sum([node.val for node in curr_level])
        